### PR TITLE
Kill off makefile-driven code-generation

### PR DESCRIPTION
**Pull request by thockin**
Get rid of the makefile-driven code-generation logic in favor of a simpler script.

The `make` rules which auto-generate some of our API stuff are incredibly baroque, and hard to maintain.  They were originally added on the assumption that we would stop checking generated files into git.  Since then we have moved away from that goal, and the worst problems with generated files have been resolved.

Reasons to kill this:
* It is slow on every build, as opposed to just being slow when running the generators.  It is even slow to calculate that there's nothing to update.
* Most development work doesn't involve changing APIs.
* It only covers about half (or less) of the generated code, and making it cover more would be even slower.
* Approximately 1 person knows how this all works.
* We have CI to make sure changes do not get merged without updating this code.
* We have corner cases where this does the WRONG thing and tracking those down is ugly and hard in perpetuity.

So this commit puts all the same logic that WAS in the Makefile.generated_files into update-codegen.sh.

I do not love this script, especially WRT sub-packages, but I am trying not to boil the ocean.  I hope to follow up with some more cleanups over time.

I have tested this manually and with the scripts and it still seems to catch errors properly.

I suggest reviewing commit-by-commit.

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by thockin** [comment]
New push leaves old rules in place but no-op.  Will have to fix out-of-repo tests then circle back to these few places.

**Comment by thockin** [comment]
This time for real, I swear.

**Comment by lavalamp** [comment]
I'm tempted to blindly LGTM anything that deletes net 1k lines of make but I will actually read this first :)

**Comment by liggitt** [comment]
swept all the commits except "Copy make's codegen logic to update-codegen.sh" and they lgtm, reading through that one now

**Comment by liggitt** [comment]
(as an aside, this took the `local-up-cluster.sh` build from 4:43 to 3:08 locally 🎉 )

**Comment by dims** [comment]
> (as an aside, this took the `local-up-cluster.sh` build from 4:43 to 3:08 locally 🎉 )

Nice savings!

File: hack/lib/util.sh [hunk 1]
```diff
@@ -798,17 +798,38 @@ function kube::util::md5() {
...
+#   declare -- X="abc"
+#   declare -a Y=([0]="a" [1]="b" [2]="c")
+#   bash: declare: Z: not found
+#   $ kube::util::read-array X < <(echo -ne "1 1\n2 2\n3 3\n")
+#   $ kube::util::read-array Y < <(echo -ne "1 1\n2 2\n3 3\n")
+#   $ kube::util::read-array Z < <(echo -ne "1 1\n2 2\n3 3\n")
+#   $ declare -p X Y Z
+#   declare -a X=([0]="1 1" [1]="2 2" [2]="3 3")
+#   declare -a Y=([0]="1 1" [1]="2 2" [2]="3 3")
+#   declare -a Z=([0]="1 1" [1]="2 2" [2]="3 3")
```
Comment (lavalamp) [inline hack/lib/util.sh:825]: It's... hard to imagine a language with a less user friendly take on arrays. (just complaining, not asking for a change, unless you can fix bash)

**Comment by thockin** [comment]
I tried to leave comments anywhere it wasn't obvious.  I ACK that "obvious" is relative.

**Review by liggitt — COMMENTED** [review]
eyeballed the codegen commit and spot checked verify scripts still catch mismatches

/lgtm

File: hack/lib/util.sh [hunk 2]
```diff
@@ -798,17 +798,38 @@ function kube::util::md5() {
...
 function kube::util::read-array {
+  if [[ -z "$1" ]]; then
+    echo "usage: $0 <varname>" >&2
+    return 1
+  fi
+  declare -p "$1" 2>/dev/null | grep -q '^declare -a' || unset -v "$1"
```
[the remark this answers is not in the document]
Comment (thockin) [reply hack/lib/util.sh:831]: declare -p prints the variable metadata on stdout, but if it is not defined on stderr.

"declare -a" means it is an array.

if grep fails to match, we unset the variable

haaaaaaaack

**Comment by thockin** [comment]
Do we have a sacrificial bash reviewer?

File: hack/lib/util.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (cblecker) [reply hack/lib/util.sh:831]: This seems like a possibly breaking change.. before we would just unset and overwrite whatever the array was.. now if the array already exists, we will append to it. Is it possible it's going to break uses were we were assuming we would just overwrite the old array no matter what?

File: build/common.sh [hunk 3]
```diff
@@ -663,12 +663,9 @@ function kube::build::sync_to_container() {
   kube::build::rsync \
     --delete \
     --filter='H /.git' \
-    --filter='- /.make/' \
     --filter='- /_tmp/' \
     --filter='- /_output/' \
     --filter='- /' \
-    --filter='H zz_generated.*' \
-    --filter='H generated.proto' \
```
Comment (cblecker) [inline build/common.sh:671]: We want to start syncing these in now?

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -14,104 +13,820 @@
...
-  pkg_dir=${nopkg_dir#apis/}
-
-
-  # skip groups that aren't being served, clients for these don't matter
-  if [[ " ${KUBE_NONSERVER_GROUP_VERSIONS} " == *" ${gv} "* ]]; then
-    continue
-  fi
-
-  GV_DIRS+=("${pkg_dir}")
+DBG_CODEGEN="${DBG_CODEGEN:-0}"
```
Comment (cblecker) [inline hack/update-codegen.sh:30]: Is there a reason we need a new debug var as opposed to the standard KUBE_VERBOSE?

File: build/common.sh (hunk 3 above)
Comment (thockin) [reply build/common.sh:671]: yes.  I admit that some of this file scares me still, but I am sure we want to sync them in.

File: hack/lib/util.sh (hunk 1 above)
Comment (thockin) [reply hack/lib/util.sh:825]: We could replace all this with python or go in a lter iteration

File: hack/lib/util.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (thockin) [reply hack/lib/util.sh:831]: We won't append, we will replace.  see the comment above wrt 'Y'.  It's super unpleasant, but we still have some folks who insist on using MacOS which has bash 3.

File: hack/update-codegen.sh (hunk 4 above)
Comment (thockin) [reply hack/update-codegen.sh:30]: I retained it from the Makefile.  As it is, setting KUBE_VERBOSE generates THOUUUUUUUUSANDS of lines of logging from the tools themselves, which isn't "helpful".

File: hack/lib/util.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (cblecker) [reply hack/lib/util.sh:831]: Okay, yes, you're right that we overwrite either way. I guess I don't understand why we don't just leave the existing "unset always" functionality then. I've tested the existing function on both bash3 and 5 and the old functionality works. It seems the unset isn't even necessary:

```bash
echo "BASH_VERSION = $BASH_VERSION"

function kube::util::read-array::old {
  if [[ -z "$1" ]]; then
    echo "usage: $0 <varname>" >&2
    return 1
  fi
  local i=0
  unset -v $1
  while IFS= read -r "$1[i++]"; do :; done
  eval "[[ \${$1[--i]} ]]" || unset "$1[i]" # ensures last element isn't empty
}

function kube::util::read-array::new {
  if [[ -z "$1" ]]; then
    echo "usage: $0 <varname>" >&2
    return 1
  fi
  declare -p "$1" 2>/dev/null | grep -q '^declare -a' || unset -v "$1"
  local i=0
  while IFS= read -r "$1[i++]"; do :; done
  eval "[[ \${$1[--i]} ]]" || unset "$1[i]" # ensures last element isn't empty
}

echo -e "---\nOLD\n---"

declare -p X

X="test"

declare -p X

kube::util::read-array::old X < <(echo -ne "1 1\n2 2\n3 3\n")

declare -p X

kube::util::read-array::old X < <(echo -ne "1 1\n2 2\n3 3\n")

declare -p X

echo -e "---\nNEW\n---"

declare -p Y

Y="test"

declare -p Y

kube::util::read-array::new Y < <(echo -ne "1 1\n2 2\n3 3\n")

declare -p Y

kube::util::read-array::new Y < <(echo -ne "1 1\n2 2\n3 3\n")

declare -p Y
```

gives me

```
 cblecker  william  ~  $  /bin/bash -c ./test.sh
BASH_VERSION = 3.2.57(1)-release
---
OLD
---
./test.sh: line 27: declare: X: not found
declare -- X="test"
declare -a X='([0]="1 1" [1]="2 2" [2]="3 3")'
declare -a X='([0]="1 1" [1]="2 2" [2]="3 3")'
---
NEW
---
./test.sh: line 43: declare: Y: not found
declare -- Y="test"
declare -a Y='([0]="1 1" [1]="2 2" [2]="3 3")'
declare -a Y='([0]="1 1" [1]="2 2" [2]="3 3")'
 cblecker  william  ~  $  /opt/homebrew/bin/bash -c ./test.sh
BASH_VERSION = 5.2.0(1)-release
---
OLD
---
./test.sh: line 27: declare: X: not found
declare -- X="test"
declare -a X=([0]="1 1" [1]="2 2" [2]="3 3")
declare -a X=([0]="1 1" [1]="2 2" [2]="3 3")
---
NEW
---
./test.sh: line 43: declare: Y: not found
declare -- Y="test"
declare -a Y=([0]="1 1" [1]="2 2" [2]="3 3")
declare -a Y=([0]="1 1" [1]="2 2" [2]="3 3")
```

So the results look identical, and the changes look safe.. other than them not really being necessary?

**Comment by thockin** [comment]
Retooled read-array with @cblecker to make it not over-write existing variables that are not arrays.

File: hack/update-codegen.sh [hunk 5]
```diff
@@ -41,15 +41,52 @@ if [[ "${DBG_CODEGEN}" == 1 ]]; then
...
+# remaining arguments.
+#
+# Example:
+#   kfind -type f -name foobar.go
+function kfind() {
+    # We want to include the "special" vendor directories which are actually
+    # part of the Kubernetes source tree (./staging/*) but we need them to be
+    # named as their ./vendor/* equivalents.  Also, we do not want all of
+    # ./vendor nor ./hack/tools/vendor nor even all of ./vendor/k8s.io.
+    find -H .                      \
```
Comment (BenTheElder) [inline hack/update-codegen.sh:54]: should we use KUBE_ROOT? or document the expectation of PWD?

File: hack/verify-flags-underscore.py [hunk 6]
```diff
@@ -62,8 +62,6 @@ def get_all_files(rootdir):
             dirs.remove('third_party')
         if '.git' in dirs:
             dirs.remove('.git')
-        if '.make' in dirs:
-            dirs.remove('.make')
         if 'BUILD' in files:
```
Comment (BenTheElder) [inline hack/verify-flags-underscore.py:65]: TODO(bentheelder): we can probably drop this BUILD logic now ...

**Review by BenTheElder — COMMENTED** [review]
/lgtm
/approve

Still too much bash logic / processing in the codegen IMHO but this is a huge simplification and should provide some nice development speed boosts for non-API changes.

Perhaps a future update is to move some of this into the generators or a thin go tool on top.

File: hack/lib/util.sh [hunk 7]
```diff
@@ -798,19 +798,53 @@ function kube::util::md5() {
...
+    echo "usage: ${FUNCNAME[0]} <varname>" >&2
+    return 1
+  fi
+  if [[ -n $(declare -p "$1" 2>/dev/null) ]]; then
+    if ! declare -p "$1" 2>/dev/null | grep -q '^declare -a'; then
+      echo "${FUNCNAME[0]}: $1 is defined but isn't an array" >&2
+      return 2
+    fi
+  fi
+  local __read_array_i=0
```
Comment (BenTheElder) [inline hack/lib/util.sh:843]:
```
In ./hack/lib/util.sh line 843:
  local __read_array_i=0
        ^------------^ SC2034: __read_array_i appears unused. Verify use (or export if used externally).

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- __read_array_i appears unused. Ve...
```

Comment (BenTheElder) [reply hack/lib/util.sh:843]: I guess this bash is too advanced for shellcheck (though, we also are probably running an outdated version).

Perhaps just add a SC2034 disable comment for now.

**Comment by thockin** [comment]
Odd that it passed before but not now.

Added a disable and pushing now.

File: hack/lib/util.sh (hunk 7 above)
Comment (thockin) [reply hack/lib/util.sh:844]: Odd that it passed before but not now.

File: hack/verify-flags-underscore.py (hunk 6 above)
Comment (thockin) [reply hack/verify-flags-underscore.py:65]: I found a few refs to bzl files, and so forth - probably can lean those up too

File: hack/update-codegen.sh (hunk 5 above)
Comment (thockin) [reply hack/update-codegen.sh:55]: Added a `cd` like most other scripts

**Comment by thockin** [comment]
sorry @cblecker I added one more change (`cd $KUBE_ROOT`)

**Comment by BenTheElder** [comment]
flake:

> Kubernetes e2e suite: [It] [sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance] 

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/112819/pull-kubernetes-conformance-kind-ga-only-parallel/1577327547459309568

/retest

**Comment by thockin** [comment]
https://user-images.githubusercontent.com/5595220/193885592-ddbad4f3-cab0-42b6-bc7f-e3f42b87639a.mp4

**Comment by BenTheElder** [comment]
This significantly reduces the build time for local e2e testing 🎉


Before (87ebae2a374):
```
$ time kind build node-image .
...
real 4m50.922s
user 1m10.921s
sys  0m33.650s

# best-case re-build with no code changes
$ time kind build node-image .
...
real 2m9.612s
user 1m4.581s
sys  0m31.496s
```

After (8db0d1a8e02, `make clean`):
```
$ time kind build node-image .
...
real 2m51.458s
user 1m10.185s
sys  0m32.387s

# best-case re-build with no code changes
$ time kind build node-image .
...
real 1m14.991s
user 1m7.726s
sys  0m32.503s
```

File: hack/verify-flags-underscore.py (hunk 6 above)
Comment (BenTheElder) [reply hack/verify-flags-underscore.py:65]: Filed https://github.com/kubernetes/kubernetes/pull/112864
Mostly found BUILD, only one .bzl 👀

File: hack/make-rules/verify.sh [hunk 8]
```diff
@@ -37,13 +37,6 @@ EXCLUDED_PATTERNS=(
   "verify-licenses.sh"           # runs in a separate job to monitor availability of the dependencies periodically
   )
 
-# Exclude generated-files-remake in certain cases, if they're running in a separate job.
-if [[ ${EXCLUDE_FILES_REMAKE:-} =~ ^[yY]$ ]]; then
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/make-rules/verify.sh:41]: https://github.com/kubernetes/test-infra/pull/27688

**Pull request merged**